x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/aggs/InternalInferenceAggregation.java

```diff
@@ -48,24 +51,44 @@ public InternalAggregation reduce(List<InternalAggregation> aggregations, Reduce
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public Object getProperty(List<String> path) {
-        Map<String, Object> resultMap = this.inferenceResult.writeResultToMap();
-
-        for (int i=0; i<path.size() -1; i++) {
-            Object value = resultMap.get(path.get(i));
-            if (value == null) {
-                throw new InvalidAggregationPathException("Cannot find an key [" + path.get(i) + "] in " + path);
-            }
-
-            if (value instanceof Map<?, ?>) {
-                resultMap = (Map<String, Object>)value;
+        if (path.isEmpty()) {
+            return this;
+        } else if (path.size() == 1) {
+            String field = path.get(0);
+            if (CommonFields.VALUE.getPreferredName().equals(field)) {
+                if (inferenceResult instanceof ClassificationInferenceResults) {
+                    return ((ClassificationInferenceResults)inferenceResult).transformedPredictedValue();
+                } else if (inferenceResult instanceof SingleValueInferenceResults) {
+                    return ((SingleValueInferenceResults)inferenceResult).value();
+                } else {
+                    return null;
+                }
```

Comment:
It might be good to add a `predictedValue()` method to the `InferenceResults` interface. That way you can access it without these castings, especially since all inference results will either have a predicted value or `null`.

Reply:
Yes the problem is `value` returns a Double and `transformedPredictedValue` returns an Object that may be a Double, String or Boolean. 

I'll take another look though as it would be nice to simplify this

Reply:
`Object getPredictedValue()` or something on the Interface could help. Fewer branches would be good :D

Reply:
I went with `Object getPredictedValue()`